allow sloped/transform shape with non-identity transformation

Qrrbrbirlbel opened the pull request:
This applies the fixes of @thinbold and @muzimuzhi also to curves and arc while not globalize xc/yc for the line function.

In my opinion, this is not a breaking change because the previous behavior was never the intended way.

**Motivation for this change**

Fixes #1058

**Checklist**

Please [signoff your commits][git-s] to explicitly state your agreement to the [Developer Certificate of Origin][DCO]. If that is not possible you may check the boxes below instead:

- [X] Code changes are licensed under [GPLv2][GPL] + [LPPLv1.3c][LPPL]
- [x] Documentation changes are licensed under [FDLv1.2][FDL]

[git-s]: https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s
[DCO]: https://developercertificate.org
[GPL]: https://www.gnu.org/licenses/gpl-2.0.html
[LPPL]: https://www.latex-project.org/lppl/lppl-1-3c.txt
[FDL]: https://www.gnu.org/licenses/fdl-1.2.html

Qrrbrbirlbel committed:
allow sloped/transform shape with non-identity transformation

This applies the fixes of thinbold and muzimuzhi also to curves and lines while not globalize xc/yc for the line.

Co-Authored-By: Yukai Chou <6376638+muzimuzhi@users.noreply.github.com>
Co-Authored-By: Thin Bold <48164818+thinbold@users.noreply.github.com>

Qrrbrbirlbel committed:
allow sloped/transform shape with non-identity transformation

This applies the fixes of thinbold and muzimuzhi also to curves and lines while not globalize xc/yc for the line.

Co-Authored-By: Yukai Chou <6376638+muzimuzhi@users.noreply.github.com>
Co-Authored-By: Thin Bold <48164818+thinbold@users.noreply.github.com>
Signed-off-by: Qrrbrbirlbel <qrrbrbirlbel+github@gmail.com>

Qrrbrbirlbel committed:
Update CHANGELOG.md

Co-Authored-By: Yukai Chou <6376638+muzimuzhi@users.noreply.github.com>
Co-Authored-By: Thin Bold <48164818+thinbold@users.noreply.github.com>

Qrrbrbirlbel committed:
Update CHANGELOG.md

Co-Authored-By: Yukai Chou <6376638+muzimuzhi@users.noreply.github.com>
Co-Authored-By: Thin Bold <48164818+thinbold@users.noreply.github.com>
Signed-off-by: Qrrbrbirlbel <qrrbrbirlbel+github@gmail.com>

muzimuzhi commented on doc/generic/pgf/CHANGELOG.md:
```diff
@@ -22,8 +23,10 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Qrrbrbirlbel
 - quark67
+- thinbold
 - Yukai Chou (muzimuzhi)
```
I was already listed on line 30. 🤔

Qrrbrbirlbel replied:
Whoops, I didn't realize that. The recent update only had your nickname so I didn't even scan past the `m` …

muzimuzhi commented on tex/generic/pgf/basiclayer/pgfcoretransformations.code.tex:
```diff
@@ -466,6 +466,10 @@
+  \let\pgf@tempaa\pgf@pt@aa
+  \let\pgf@tempab\pgf@pt@ab
+  \let\pgf@tempba\pgf@pt@ba
+  \let\pgf@tempbb\pgf@pt@bb
```
The three pair of additions to `\pgftransformlineattime`, `\pgftransformarcaxesattime`, and `\pgftransformcurveattime` are so similar that maybe two new macros abstracting them correspondingly can be provided.

Qrrbrbirlbel replied:
Good idea, unfortunately `\pgftransformlineattime` uses `\pgf@xc`/`\pgf@yc` and `\pgf@x`/`\pgf@y` isn't save to use inbetween. I would just refactor that so that it uses x/y instead of xc/yc. (Too big of a change?)

That said, there's a lot more that they have in common.

muzimuzhi reviewed (commented):
Only reviewed the styling. I already forgot my own patches. :)

A new test is needed, though it could be added after #1116 is handled first, which has blocked all PRs more than just typo fixes for a while.

Qrrbrbirlbel commented:
> A new test is needed, though it could be added after #1116 is handled first, which has blocked all PRs more than just typo fixes for a while.

I do know that there's that big #1116 delaying everything else (but I got impatient because it really gets in the way oftentimes – `xscale` and `yscale` as in #1287's source is probably rare [and avoidable by scaling via `x` and `y` keys] – but just rotating, a very common transformation, is also broken this way).

I do not know much about “a new test”, though, and what is required from me.

Qrrbrbirlbel committed:
Refactored \pgftransform…attime

Qrrbrbirlbel committed:
Refactored \pgftransform…attime

Signed-off-by: Qrrbrbirlbel <qrrbrbirlbel+github@gmail.com>

Signoff

Signed-off-by: Qrrbrbirlbel <qrrbrbirlbel+github@gmail.com>

Qrrbrbirlbel commented:
Here's a `\pgftransform@commonattime` macro that combines all the common parts of the `\pgftransform…attime` commands:

- `#1` needs to setup `\pgf@x` and `\pgf@y` for the position, i.e. the `\pgftransformshift`, which is done by the corresponding `\pgfpoint…attime`.
- `#2` needs to setup `\pgf@x` and `\pgf@y` to be the tangent vector.

The, probably, legacy code of `\pgftransformlineattime` calculated the tangent again; this is now extracted from `\pgfpointlineattime` – just as it is with the others.

Here's a testfile with many transformations and all three types of paths (following the example of #1287).
It also contains an adjustment to the `\tikz@timer@line` timer that would allow distances for the `pos` key, which then uses a new `\pgftransformlineatdistance` macro whose definition is now just as easy as for the `…attime` macros.

(And, as I now realize, this also shows that the name <code>\pgftransform@commonat<b>time</b></code> isn't really correct, either.)

I've also adjusted the examples in the source since they either used deprecated macros (`\pgfxy`) or were simply wrong.

```latex
\documentclass[tikz]{standalone}
\begin{document}
\tikzset{
  every picture/.append style={
    xscale=2,    yscale=3,
    xshift=40pt, yshift=50pt,
    xslant=.2,   yslant=-.3,
    rotate=-35,
    nodes=above,
    execute at begin picture={
      \draw[help lines, step=1] (0,0) grid (7,3);
      \draw[thick, <->] (0,1) |- (1,0);}}}
%\makeatletter
%\let\iftikz@time@isdistance\iffalse
%\tikzset{
%  pos/.code=\edef\tikz@time{#1}\pgfutil@ifxempty\tikz@time{}{%
%              \pgfmathsetmacro\tikz@time{\tikz@time}%
%              \let\iftikz@time@isdistance\ifpgfmathunitsdeclared}}
%\def\tikz@timer@line{%
%  \iftikz@time@isdistance
%    \pgftransformlineatdistance{\tikz@time pt}{\tikz@timer@start}{\tikz@timer@end}%
%  \else
%    \pgftransformlineattime{\tikz@time}{\tikz@timer@start}{\tikz@timer@end}%
%  \fi}%
%\def\pgftransformlineatdistance#1#2#3{%
%  \pgftransform@commonattime{\pgfpointlineatdistance{#1}{#2}{#3}}{\pgf@x=\pgf@xa\pgf@y=\pgf@ya}}
%\makeatother
\begin{tikzpicture}
\draw (0, 0) -- node[sloped                 ] {\texttt{sloped} only}                         +(2, 2);
\draw (2, 0) -- node[        transform shape] {\texttt{transform shape} only}                +(2, 2);
\draw (4, 0) -- node[sloped, transform shape] {\texttt{sloped} and \texttt{transform shape}} +(2, 2);
\end{tikzpicture}
%\begin{tikzpicture}[pos=2cm]
%\draw (0, 0) -- node[sloped                 ] {\texttt{sloped} only}                         +(2, 2);
%\draw (2, 0) -- node[        transform shape] {\texttt{transform shape} only}                +(2, 2);
%\draw (4, 0) -- node[sloped, transform shape] {\texttt{sloped} and \texttt{transform shape}} +(2, 2);
%\end{tikzpicture}
\begin{tikzpicture}
\draw (0, 0) to[bend left] node[sloped                 ] {\texttt{sloped} only}                         +(2, 2);
\draw (2, 0) to[bend left] node[        transform shape] {\texttt{transform shape} only}                +(2, 2);
\draw (4, 0) to[bend left] node[sloped, transform shape] {\texttt{sloped} and \texttt{transform shape}} +(2, 2);
\end{tikzpicture}
\begin{tikzpicture}[start angle=180, delta angle=-120, x radius=1.5, y radius=.75, nodes=midway]
\draw (0, 0) arc[] node[sloped                 ] {\texttt{sloped} only}                        ;
\draw (2, 0) arc[] node[        transform shape] {\texttt{transform shape} only}               ;
\draw (4, 0) arc[] node[sloped, transform shape] {\texttt{sloped} and \texttt{transform shape}};
\end{tikzpicture}
\end{document}
```

Qrrbrbirlbel committed:
Merge branch 'master' of https://github.com/Qrrbrbirlbel/tikz-sloped

Qrrbrbirlbel committed:
Merge branch 'master' into master
